src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/TraceReporter.kt

```diff
@@ -79,14 +70,30 @@ private fun splitToColumns(nThreads: Int, traceRepresentation: List<TraceEventRe
  */
 private fun constructTraceGraph(scenario: ExecutionScenario, results: ExecutionResult?, trace: Trace): TraceNode? {
     val tracePoints = trace.trace
+    // remap scenario actors: put init/post part into first thread
+    val remappedScenario = Array(scenario.threads) { i ->
```

Comment:
I see a lot of logic here and there related to the init/post parts. Consider incorporating init/post parts into the first thread scenario part (as you depict when printing an error message), making the strategy responsible for executing init/post parts in the first thread at the beginning/end of the scenario. Would it simplify the overall design?

Reply:
I don't see why you must modify the `ExecutionScenario` class.

Reply:
I've tried to address this problem and simplify thread and actors enumeration by putting all the related logic in `ExecutionScenario` and `ExecutionResult` classes.

There is still some room for improvement. For example, I was not able to refactor happens-before clock calculation logic --- it just too complicated and convoluted (because of bytecode generation). I think we first need to refactor `TestThreadExecutionGenerator` class and convert as much code as possible from bytecode generation to regular Kotlin code. I'll open an issue for this problem. 

For now I come up with an intermediate solution that just rebuilds clocks inside `ExecutionResult` class.

Reply:
Is this discussion up-to-date?